### ISSUE TITLE
Wesnoth uses 100% CPU unless OMP_WAIT_POLICY is manually set

### ISSUE THREAD

**Issue by marmistrz** [wesnoth/wesnoth#4130]
TL;DR: The workaround is:
```
export OMP_WAIT_POLICY=passive
wesnoth
```

This is different from #1474, because that issue is not OpenMP-related.
This issue was already mentioned at the [wesnoth forums](https://forums.wesnoth.org/viewtopic.php?f=4&t=48666&sid=c968877fbeabf499c194a35012faa41f&start=15), but is missing from the bugtracker. It looks like some of the Windows scripts (e.g. cwesnoth.cmd) already take this variable into account, but this is not the case on Linux.

The [gcc documentation](https://gcc.gnu.org/onlinedocs/libgomp/OMP_005fWAIT_005fPOLICY.html#OMP_005fWAIT_005fPOLICY) mentions that

> Specifies whether waiting threads should be active or passive. If the value is PASSIVE, waiting threads should not consume CPU power while waiting; while the value is ACTIVE specifies that they should. If undefined, threads wait actively for a short time before waiting passively. 

In other words, it looks like wesnoth has a lot of frequent but light computation, which makes the OpenMP threads spinlock all the time. On the other hand, why would idling in a hot-seat match require that much computation?

**Comment by gfgtdf** [comment]
Iirc we removed the omp code in master so this should not longer happen in 1.15

**Comment by sevu** [comment]
It probably was reverted with the new master.

**Comment by soliton-** [comment]
Indeed there are still parts left in src/display.cpp in master which should presumably just be removed.

1.14 already sets `OMP_WAIT_POLICY` to `PASSIVE` and restarts itself if it isn't set.

https://github.com/wesnoth/wesnoth/blob/3771ac0f10ff3de34a66473c9f4ff52af2191f1e/src/wesnoth.cpp#L1057-L1081

Not sure if that happens in every situation that openmp could be enabled though.

**Comment by marmistrz** [comment]
@soliton- I'm experiencing the issue on v1.14.7, on Arch Linux.

**Comment by soliton-** [comment]
Well, perhaps you can figure out what's going on in more detail. The above code means that for Arch Linux the part between line 1070 and 1075 is run which sets `OMP_WAIT_POLICY` to `PASSIVE` and restarts itself unless `OMP_WAIT_POLICY` is already set to something.

Is `OMP_WAIT_POLICY` set to something before starting wesnoth? If `echo ${OMP_WAIT_POLICY+set}` outputs set then it is.

You could also run wesnoth under strace: `strace -ff -e process wesnoth` and see if you can verify that it does or doesn't restart itself.

**Comment by marmistrz** [comment]
Notice that wesnoth doesn't check if `execv` syscall error code. `strace` shows that it fails:
```
execve("wesnoth", ["wesnoth"], 0x564cea84fc30 /* 52 vars */) = -1 ENOENT (No such file or directory)
```
OTOH, the `execve` syscall succeeds if the game is launched using the full path, i.e. as `/usr/bin/wesnoth`.

**Comment by soliton-** [comment]
So wesnoth is not able to restart itself. We should perhaps abort at that point or at least show an error message. Can you show more of the strace output? What does the first execve look like?

**Comment by jostephd** [comment]
execve doesn't do a path search. `execve("/usr/games/wesnoth", ...)` (or whatever the path to the wesnoth binary is) would be valid; `execvpe("wesnoth", ...)` would be valid; `execve("wesnoth")` is not. (But execvpe is a GNU extension, it's not portable.)

**Comment by soliton-** [comment]
That is why I want to see the first `execve` to see how wesnoth was started initially. Specifically what's in `argv[0]`. Though it's likely going to be just "wesnoth" so it seems this restart code is not particularly useful to begin with. Probably was only tested with the wesnoth binary in the current working directory.

**Comment by soliton-** [comment]
A trivial fix is probably to just use `execvp` instead of `execv`.

**Comment by GregoryLundberg** [comment]
Seems to me the better fix would be to backport the removal of OpenMP and make the problem go away completely.

**Comment by marmistrz** [comment]
It's how the calls look like from strace:
```
execve("/usr/bin/wesnoth", ["wesnoth"], 0x7ffdedeefb08 /* 51 vars */) = 0
execve("wesnoth", ["wesnoth"], 0x55741da1bc30 /* 52 vars */) = -1 ENOENT (No such file or directory)
```

**Comment by Vultraz** [comment]
OpenMP removed on 1.14, remnants cleaned up from master.
